akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2247,8 +2249,18 @@ private[stream] object Collect {
         })
 
         subOutlet.setHandler(new OutHandler {
-          override def onPull(): Unit = {
-            pull(in)
+          override def onPull(): Unit = unpushedFirstElement match {
+            case OptionVal.Some(firstElem) =>
+              // the demand can be satisfied right away by the cached element
+              unpushedFirstElement = OptionVal.none
+              subOutlet.push(firstElem)
+              // in.onUpstreamFinished was not propagated if it arrived before the cached element was pushed
+              // -> check if the completion must be propagated now
+              if (isClosed(in)) {
+                subOutlet.complete()
+                maybeCompleteStage()
+              }
+            case OptionVal.None => pull(in)
```

Comment:
this looks like a reimplementation of 'prepend' and I'm pretty sure it's not necessary here.
I propose a much simpler approach: when subOutlet is pulled (by prepended internal flow) it delegates to 'in', when 'in' is completed/failed it delegates to subOutlet. remember the internal flow has its own behaviour for dealing with completion (i.e. concat, orElse, prepend...) and errors (i.e. recover) so you have to make sure to delegate events to the internal prepended flow so it can take care of them.